### ISSUE TITLE
about MACs

### ISSUE THREAD

**Issue by Sunshine-Ye** [zhuogege1943/dgc#1]
thanks the authors for this promising work!

In the paper, the convolution in the ith head is conducted with only the selected emphasized channels and  the corresponding ﬁlters:
![image](https://user-images.githubusercontent.com/32639674/87289161-dde28980-c52e-11ea-8379-bc0ccfb13a16.png)

In the code, the unselected channels is also computed with the corresponding ﬁlters, instead the unselected channels is masked with 0:
![image](https://user-images.githubusercontent.com/32639674/87290191-441bdc00-c530-11ea-90e1-8334a9d2e35d.png)
![image](https://user-images.githubusercontent.com/32639674/87290293-63b30480-c530-11ea-9d19-116c6d48b10a.png)

I want know if the actual MACs or flops or computation-cost are effectively reduced in the pytorch implementation?

**Comment by hellozhuo** [comment]
Yes you are right, the prunned input channels are simply assigned as 0s for training convenience, therefore the MACs are not reduced in this implementation. In order to achieve the theoretical computation reduction in the testing phase, as discussed at the end of the paper, a dynamic version of "index layer" should be applied, which is supposed to gather the selected channels to form the input for the group convolution.  We also suggest you to refer to the the discussion of index layer in [condensenet](https://arxiv.org/abs/1711.09224) and [FLGC](https://arxiv.org/abs/1904.00346).

**Comment by Sunshine-Ye** [comment]
how to achieve a dynamic version of "index layer"? i mean, how to realize it in pytorch? the index layer in condensenet and FLGC is "static" while it  is "dynamic" in DGC, how to achieve the theoretical computation reduction in the testing phase?
look forward for your reply,  an example with pytorch code may be better.

**Comment by hellozhuo** [comment]
@Sunshine-Ye  In Pytroch framework, you could use something like [index select](https://discuss.pytorch.org/t/batched-index-select/9115) once you get the indices of the selected channels.  This works when batch size is 1. For batch size > 1, it is true a problem at least under the current Pytorch framework, because individual sample has its different filters corresponding to the selected channels.  We are also looking forward to new solutions.
